compiler/rustc_trait_selection/src/traits/misc.rs

```diff
@@ -47,12 +56,75 @@ pub fn type_allowed_to_implement_copy<'tcx>(
         | ty::Ref(_, _, hir::Mutability::Not)
         | ty::Array(..) => return Ok(()),
 
-        ty::Adt(adt, substs) => (adt, substs),
+        &ty::Adt(adt, substs) => (adt, substs),
 
         _ => return Err(CopyImplementationError::NotAnAdt),
     };
 
-    let copy_def_id = tcx.require_lang_item(hir::LangItem::Copy, Some(parent_cause.span));
+    all_fields_implement_trait(
+        tcx,
+        param_env,
+        self_type,
+        adt,
+        substs,
+        parent_cause,
+        hir::LangItem::Copy,
+    )
+    .map_err(CopyImplementationError::InfrigingFields)?;
+
+    if adt.has_dtor(tcx) {
+        return Err(CopyImplementationError::HasDestructor);
+    }
+
+    Ok(())
+}
+
+/// Checks that the fields of the type (an ADT) all implement `ConstParamTy`.
+///
+/// If fields don't implement `ConstParamTy`, return an error containing a list of
+/// those violating fields.
+///
+/// If it's not an ADT, int ty, `bool` or `char`, returns `Err(NotAnAdtOrBuiltinAllowed)`.
+pub fn type_allowed_to_implement_const_param_ty<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    self_type: Ty<'tcx>,
+    parent_cause: ObligationCause<'tcx>,
+) -> Result<(), ConstParamTyImplementationError<'tcx>> {
+    let (adt, substs) = match self_type.kind() {
+        // `core` provides these impls.
+        ty::Uint(_) | ty::Int(_) | ty::Bool | ty::Char => return Ok(()),
+
+        &ty::Adt(adt, substs) => (adt, substs),
+
+        _ => return Err(ConstParamTyImplementationError::NotAnAdtOrBuiltinAllowed),
```

Comment:
Looking through `TyKind` I think we need more special cases here :sob:, some of them are going to need to evaluate obligations to see if its allowed or not though :thinking: 

- `ty::Str` should be usable and probably require `[u8]: ConstParamTy` to act as if `Str` was `struct Str([u8])`. 
  - Should add an impl for `impl ConstParamTy for str` in core
- `ty::Array` should be usable and require `T: ConstParamTy`
  - Should add an impl for `impl<T: ConstParamTy, const N: usize> ConstParamTy for [T; N]` in core
- `ty::Slice` should be usable and require `T: ConstParamTy`
  - Should add an impl for `impl<T: ConstParamTy> ConstParamTy for [T]` in core
- `ty::Ref` when its `Mutability::Not` and require `T: ConstParamTy`
  - Should add an impl for `impl<T: ConstParamTy> ConstParamTy for &'_ T ` in core 
- `ty::Tuple` should be usable and require `T: ConstParamTy` for all the elements
  - Should add an impl in core via a macro? How does std do its current impls for tuples xd

I think eventually we want some logic around `ty::FnDef` and `ty::Closure` but that is likely going to require builtin impl candidates, and without `feature(generic_const_parameter_types)` you can't even write an `adt_const_param` that has one of those types. Could you add a `FIXME(generic_const_parameter_types)` here to make `ty::FnDef`/`ty::Closure` work?

Reply:
> `generic_const_parameter_types`
Is it okay that this feature does not exist?

Reply:
yea its not yet implemented lol but that will be what i call the feature gate when its done :sweat_smile:

Reply:
this match needs to actually evaluate obligations checking the things i said should be required so that you cant even write an impl like `impl<T> ConstParamTy for &T {}`

Reply:
Wouldn't it duplicate the bounds we have anyway in the `core`? I don't really see the reason why we need to also check them here (`type_allowed_to_implement_copy` doesn't, as an example).

Reply:
My point is: we already depend on std being correct in a myriad of ways, for all intents and purposes from the perspective of the compiler `std` **must** be correct.

Reply:
I don't see why we should rely on std being correct with regards to this, its easily checkable and no where is it clear that implementing `Copy` or `ConstParamTy` is actually unsafe inside of std